src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -684,19 +685,22 @@ internal sealed override bool IsInterface
             }
         }
 
-        internal override ManagedKind ManagedKind
+        internal override ManagedKind GetManagedKind(ref HashSet<DiagnosticInfo> useSiteDiagnostics)
         {
-            get
+            var managedKind = _flags.ManagedKind;
+            if (managedKind == ManagedKind.Unknown)
             {
-                var managedKind = _flags.ManagedKind;
-                if (managedKind == ManagedKind.Unknown)
-                {
-                    var baseKind = base.ManagedKind;
-                    _flags.SetManagedKind(baseKind);
-                    return baseKind;
-                }
-                return managedKind;
+                managedKind = base.GetManagedKind(ref _managedKindUseSiteDiagnostics);
```

Comment:
>_managedKindUseSiteDiagnostics [](start = 54, length = 30)

It feels like there might be a race here. I am also not sure if HashSet.Add is thread safe operation. I would collect in a local set and then interlocked initialize an immutable array field from it.

Reply:
Thanks, that way is much better and fits with established patterns.